ansible_sdk/_util/metrics.py

```diff
@@ -0,0 +1,122 @@
+# Copyright: Ansible Project
+# Apache License 2.0 (see LICENSE or https://www.apache.org/licenses/LICENSE-2.0)
+
+import os
+import uuid
+
+from datetime import datetime
+
+import ansible_sdk._util.dataclass_compat as dataclasses
+from ansible_sdk.model.job_status import AnsibleJobStatus
+from ansible_sdk._aiocompat.csv_async import CSVAsync
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class AnsiblePlaybookStats:
+    job_id: str
+    job_type: str
+    started: str
+    finished: str
+    job_state: str
+    hosts_ok: str
+    hosts_changed: str
+    hosts_skipped: str
+    hosts_failed: str
+    hosts_unreachable: str
+    task_count: str
+    task_duration: str
+
+
+class MetricsCalc:
+    def __init__(self):
+        pass
+
+    async def collect_metrics(self, status_obj: AnsibleJobStatus):
+        metrics_data = {
+            "task_counter": 0,
+            "task_data": {},
+            "role_data": {},
+        }
+
+        if not status_obj._job_def.metrics_output_path:
+            return
+
+        # Create job.csv
+        job_id = uuid.uuid4()
+        job_csv_filename = os.path.join(
+            status_obj._job_def.metrics_output_path,
+            "job_%s.csv" % job_id,
+        )
+        job_csv_fh = open(job_csv_filename, "w+")
+        headers = [x.name for x in dataclasses.fields(AnsiblePlaybookStats)]
+        job_csv_writer = CSVAsync(job_csv_fh, headers, restval="NULL")
+        await job_csv_writer.writeheader_async()
+
+        # Create modules.csv
+        module_csv_filename = os.path.join(
+            status_obj._job_def.metrics_output_path,
+            "modules_%s.csv" % job_id,
+        )
+        module_csv_fh = open(module_csv_filename, "w+")
```

Comment:
Same as above; I'd suggest reordering things so there's a single `with` statement with multiple file handle context managers wrapping all of this.